bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/ComputeAndValidateHashFromBigtableDoFn.java

```diff
@@ -0,0 +1,232 @@
+import org.apache.hadoop.hbase.mapreduce.BigtableTableHashAccessor.BigtableResultHasher;
+
+/**
+ * A {@link DoFn} that takes a row range and hash from HBase and validates the hash from rows read
+ * from Cloud Bigtable.
+ */
+@InternalApi
+class ComputeAndValidateHashFromBigtableDoFn
+    extends AbstractCloudBigtableTableDoFn<KV<String, Iterable<List<RangeHash>>>, RangeHash> {
+
+  private static final long serialVersionUID = 2349094L;
+  private final ValueProvider<String> tableName;
+  private final ValueProvider<String> projectId;
+  private final ValueProvider<String> sourceHashDir;
+
+  private final TableHashWrapperFactory tableHashWrapperFactory;
+
+  // Counter for reporting matching and mismatching ranges. Names are similar to HBase sync-table
+  // job.
+  private final Counter matches = Metrics.counter("cbt-dataflow-validate", "ranges_matched");
+  private final Counter mismatches = Metrics.counter("cbt-dataflow-validate", "ranges_not_matched");
+
+  public ComputeAndValidateHashFromBigtableDoFn(SyncTableOptions options) {
+    super(TemplateUtils.BuildSyncTableConfig(options));
+    this.tableName = options.getBigtableTableId();
+    // Create a local copy of ValueProviders, PipelineOptions are not serializable.
+    projectId = options.getBigtableProject();
+    sourceHashDir = options.getHashTableOutputDir();
+    tableHashWrapperFactory = new TableHashWrapperFactory();
+  }
+
+  @VisibleForTesting
+  ComputeAndValidateHashFromBigtableDoFn(
+      CloudBigtableConfiguration config,
+      ValueProvider<String> tableName,
+      ValueProvider<String> projectId,
+      ValueProvider<String> sourceHashDir,
+      TableHashWrapperFactory factory) {
+    super(config);
+    this.tableName = tableName;
+    this.tableHashWrapperFactory = factory;
+    this.sourceHashDir = projectId;
+    this.projectId = sourceHashDir;
+  }
+
+  @ProcessElement
+  public void processElement(ProcessContext context) throws Exception {
+    // BufferedHadoopHashTableSource generates only 1 item per groupby key, but iterate just in
+    // case.
+    for (List<RangeHash> rangeHashes : context.element().getValue()) {
+      if (rangeHashes.isEmpty()) {
+        // No rows ranges found, return;
+        return;
+      }
+
+      ImmutableBytesWritable rangeStartInclusive = rangeHashes.get(0).startInclusive;
+      ImmutableBytesWritable rangeEndExclusive =
+          rangeHashes.get(rangeHashes.size() - 1).stopExclusive;
+
+      BigtableResultHasher resultHasher = new BigtableResultHasher();
+      resultHasher.startBatch(rangeStartInclusive);
+
+      // Since all the row-ranges are sorted in HashTable's data files, 1 big scan can be used
+      // to read all the row ranges. Parallelism is achieved by splitting the HashTable's data
+      // files into smaller bundle of row-ranges in GroupBy.
+      ResultScanner scanner =
+          createBigtableScan(rangeStartInclusive.copyBytes(), rangeEndExclusive.copyBytes());
```

Comment:
why copyBytes instead of get()?

Reply:
get() returns the full byte[] and the caller is responsible to prune it to the bounds covered by this object. 

```
 public byte[] get() {
    if (this.bytes == null) {
      throw new IllegalStateException("Uninitialiized. Null constructor called w/o accompaying readFields invocation");
    } else {
      return this.bytes;
    }
  }


  public byte[] copyBytes() {
    return Arrays.copyOfRange(this.bytes, this.offset, this.offset + this.length);
  }
```

Reply:
Thats a bit annoying. How would you feel about adding a helper to avoid copies in most cases:

```java
byte[] unwrapBytes(ImmutableBytesWriteable w) {
  byte[] innerBytes = w.getBytes();
  if (w.getOffset() == 0 && w.getLength() == innerBytes.length) {
    return innerBytes;
  } else {
     return w.copyBytes();
  }
}
```